test/framework/ownerreference_helpers.go

```diff
@@ -0,0 +1,308 @@
+
+	// Check that the ownerReferences are as expected after additional reconciliations.
+	AssertOwnerReferences(namespace, proxy.GetKubeconfigPath(), assertFuncs...)
+}
+
+func AssertOwnerReferences(namespace, kubeconfigPath string, assertFuncs ...map[string]func(reference []metav1.OwnerReference) error) {
+	allAssertFuncs := map[string]func(reference []metav1.OwnerReference) error{}
+	for _, m := range assertFuncs {
+		for k, v := range m {
+			allAssertFuncs[k] = v
+		}
+	}
+	Eventually(func() error {
+		graph, err := clusterctlcluster.GetOwnerGraph(namespace, kubeconfigPath)
+		Expect(err).To(BeNil())
+		allErrs := []error{}
+		for _, v := range graph {
+			if _, ok := allAssertFuncs[v.Object.Kind]; !ok {
+				allErrs = append(allErrs, fmt.Errorf("kind %s does not have an associated ownerRef assertion function", v.Object.Kind))
+				continue
+			}
+			if err := allAssertFuncs[v.Object.Kind](v.Owners); err != nil {
+				allErrs = append(allErrs, errors.Wrapf(err, "Unexpected ownerReferences for %s/%s", v.Object.Kind, v.Object.Name))
+			}
+		}
+		var errStrings string
+		for _, err := range allErrs {
+			errStrings = fmt.Sprintf("%s\n%s", errStrings, err.Error())
+		}
+		return kerrors.NewAggregate(allErrs)
+	}).Should(Succeed())
+}
+
+// Kind and GVK for types in the core API package.
+var (
+	clusterClassKind       = "ClusterClass"
+	clusterKind            = "Cluster"
+	machineKind            = "Machine"
+	machineSetKind         = "MachineSet"
+	machineDeploymentKind  = "MachineDeployment"
+	machineHealthCheckKind = "MachineHealthCheck"
+
+	clusterClassGVK      = clusterv1.GroupVersion.WithKind(clusterClassKind)
+	clusterGVK           = clusterv1.GroupVersion.WithKind(clusterKind)
+	machineDeploymentGVK = clusterv1.GroupVersion.WithKind(machineDeploymentKind)
+	machineSetGVK        = clusterv1.GroupVersion.WithKind(machineSetKind)
+	machineGVK           = clusterv1.GroupVersion.WithKind(machineKind)
+)
+
+// CoreTypeOwnerReferenceAssertion maps Cluster API core types to functions which return an error if the passed
+// OwnerReferences aren't as expected.
+var CoreTypeOwnerReferenceAssertion = map[string]func([]metav1.OwnerReference) error{
+	clusterClassKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{})
+	},
+	clusterKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{})
+	},
+	machineDeploymentKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{clusterGVK})
+	},
+	machineSetKind: func(owners []metav1.OwnerReference) error {
+		return hasExactOwnersByGVK(owners, []schema.GroupVersionKind{machineDeploymentGVK})
+	},
+	machineKind: func(owners []metav1.OwnerReference) error {
+		return hasOneOfExactOwnersByGVK(owners, []schema.GroupVersionKind{machineSetGVK}, []schema.GroupVersionKind{kubeadmControlPlaneGVK})
```

Comment:
Let's document why we are using hasOneOfExactOwnersByGVK or most specifically let's document in which case we have one owner ref, in which case the other.
Same for the other kinds